File: nav_pkg/scripts/nav_client.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
import rospy
import actionlib
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal
import time

def simple_action(x, y, w):
    goal.target_pose.pose.position.x = x
    goal.target_pose.pose.position.y = y
    goal.target_pose.pose.orientation.w = w
    ac.send_goal(goal)
    ac.send_goal(goal)
    rospy.loginfo("Sent goal")
    ac.wait_for_result()
    if ac.get_state() == actionlib.GoalStatus.SUCCEEDED:
        rospy.loginfo("Goal succeeded!")
    else:
        rospy.loginfo("Goal failed with error code: " + str(ac.get_state()))


if __name__ == "__main__":
    rospy.init_node("nav_client")
    ac = actionlib.SimpleActionClient("move_base", MoveBaseAction)
    ac.wait_for_server()  # 等待服务
    goal = MoveBaseGoal()  # 目标点
    goal.target_pose.header.frame_id = "map"
    goal.target_pose.header.stamp = rospy.Time.now()
    goal.target_pose.pose.position.x = -3.0
    goal.target_pose.pose.position.y = 2.0
    goal.target_pose.pose.orientation.w = 1.0
    ac.send_goal(goal)
    ac.send_goal(goal)
    rospy.loginfo("Sent goal")
    ac.wait_for_result()
    if ac.get_state() == actionlib.GoalStatus.SUCCEEDED:
        rospy.loginfo("Goal succeeded!")
    else:
        rospy.loginfo("Goal failed with error code: " + str(ac.get_state()))
    simple_action(0.0, 0.0, 1.0)
    time.sleep(5)
    simple_action(2.0, -2.0, 1.0)
    time.sleep(5)
    simple_action(0.0, 0.0, 1.0)

File: 6-robot/nav_pkg/src/nav_client.cpp
#include <ros/ros.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <actionlib/client/simple_action_client.h>
#include <string>
#include <cstdlib> // for atof

typedef actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> MoveBaseClient;

// 声明为全局变量以便在函数间共享
MoveBaseClient *ac_ptr = nullptr;

void sample_goal(double x_goal, double y_goal)
{
  if (!ac_ptr || !ac_ptr->isServerConnected())
  {
    ROS_ERROR("Action client not initialized or server not available");
    return;
  }

  ROS_INFO("Sending goal to (%.2f, %.2f)", x_goal, y_goal);

  move_base_msgs::MoveBaseGoal goal;
  goal.target_pose.header.frame_id = "map";
  goal.target_pose.header.stamp = ros::Time::now();
  goal.target_pose.pose.position.x = x_goal;
  goal.target_pose.pose.position.y = y_goal;
  goal.target_pose.pose.orientation.w = 1.0; // 默认朝向

  ac_ptr->sendGoal(goal);

  bool finished_before_timeout = ac_ptr->waitForResult(ros::Duration(30.0)); // 30秒超时

  if (finished_before_timeout)
  {
    actionlib::SimpleClientGoalState state = ac_ptr->getState();
    if (state == actionlib::SimpleClientGoalState::SUCCEEDED)
    {
      ROS_INFO("Mission complete!");
    }
    else
    {
      ROS_WARN("Mission failed: %s", state.toString().c_str());
    }
  }
  else
  {
    ROS_WARN("Action did not complete before timeout");
  }
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "demo_simple_goal");

  // 初始化action client
  MoveBaseClient ac("move_base", true);
  ac_ptr = &ac;

  // 等待action server
  ROS_INFO("Waiting for move_base server...");
  if (!ac.waitForServer(ros::Duration(5.0)))
  {
    ROS_ERROR("Could not connect to move_base server");
    return 1;
  }

  // 发送目标点
  sample_goal(1.0, 0.0);
  sample_goal(0.0, 1.0);
  sample_goal(0.0, 0.0);

  // 清理资源
  ac_ptr = nullptr;
  ros::shutdown();
  return 0;
}